File: packages/flutter_mozilla_components/android/src/test/kotlin/app/mellow/browser/flutter_mozilla_components/startup/MaintenanceScannerTest.kt
/*
 * Copyright (c) 2024-2026 Fabian Freund.
 *
 * This file is part of WebLibre
 * (see https://weblibre.eu).
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 */
package eu.weblibre.flutter_mozilla_components.startup

import java.io.File
import java.nio.file.Files
import kotlin.test.AfterTest
import kotlin.test.BeforeTest
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertTrue
import org.json.JSONArray
import org.json.JSONObject

private const val PROFILE_ID = "0199a0b1-1111-7111-8111-111111111111"

/** Mirrors `apps/weblibre/test/core/startup/maintenance_scanner_test.dart`. */
class MaintenanceScannerTest {

    private lateinit var filesDir: File
    private lateinit var paths: StartupPaths

    @BeforeTest
    fun setUp() {
        filesDir = Files.createTempDirectory("weblibre_maintenance").toFile()
        paths = StartupPaths(filesDir)
        paths.profilesDir.mkdirs()
        paths.maintenanceJournalsDir.mkdirs()
        paths.maintenanceRestoreDir.mkdirs()
    }

    @AfterTest
    fun tearDown() {
        filesDir.deleteRecursively()
    }

    private fun writeJournal(taskId: String, kind: String, phase: String) {
        paths.journalFile(taskId).writeText(
            JSONObject()
                .put("version", MAINTENANCE_JOURNAL_VERSION)
                .put("taskId", taskId)
                .put("kind", kind)
                .put("phase", phase)
                .put("targetProfileId", PROFILE_ID)
                .put("participants", JSONArray())
                .put("updatedAt", "2026-08-18T00:00:00.000Z")
                .toString(),
        )
    }

    private fun configWith(vararg tasks: Pair<String, MaintenanceTaskState>) = StartupConfig(
        pendingTasks = tasks.map { (id, state) ->
            MaintenanceTaskSummary(
                id = id,
                actionId = MaintenanceAction.BACKUP.id,
                stateId = state.id,
                profileId = PROFILE_ID,
                profileName = "Default",
            )
        },
    )

    @Test
    fun anEmptyWorkspaceProducesNoReservation() {
        val scan = MaintenanceScan.scan(paths)

        assertFalse(scan.hasDurableEvidence)
        val reservation = MaintenanceReservation.resolve(StartupConfig.DEFAULTS, scan)
        assertFalse(reservation.required)
        assertFalse(reservation.recoveryRequired)
    }

    @Test
    fun aCompletedJournalAloneDoesNotReserveMaintenance() {
        writeJournal("task-1", MaintenanceJournalKind.DELETE.id, DeletePhase.COMPLETED.id)

        val scan = MaintenanceScan.scan(paths)
        assertEquals(1, scan.journals.size)
        assertTrue(scan.incompleteJournals.isEmpty())
        assertFalse(scan.hasDurableEvidence)
    }

    @Test
    fun anIncompleteJournalReservesMaintenanceAndDemandsRecovery() {
        writeJournal("task-1", MaintenanceJournalKind.RESTORE.id, RestorePhase.OLD_MOVED.id)

        val reservation = MaintenanceReservation.resolve(
            StartupConfig.DEFAULTS,
            MaintenanceScan.scan(paths),
        )

        assertTrue(reservation.required)
        assertTrue(reservation.recoveryRequired)
        assertEquals("task-1", reservation.taskId)
        assertEquals(setOf(PROFILE_ID), reservation.scan.profilesUnderMaintenance)
    }

    @Test
    fun anUnreadableJournalIsReportedNotQuarantinedAway() {
        paths.journalFile("task-1").writeText("{ not json")

        val scan = MaintenanceScan.scan(paths)

        assertEquals(1, scan.unreadableJournals.size)
        assertTrue(scan.hasDurableEvidence)
        // The evidence must survive the scan, or the next process would boot the
        // possibly half-restored profile.
        assertTrue(paths.journalFile("task-1").exists())
    }

    @Test
    fun aJournalWithoutIdentifyingFieldsCountsAsUnreadable() {
        paths.journalFile("task-1").writeText("""{"phase":"staged"}""")

        val scan = MaintenanceScan.scan(paths)
        assertEquals(1, scan.unreadableJournals.size)
        assertTrue(scan.journals.isEmpty())
    }

    @Test
    fun restoreArtifactsReserveMaintenanceEvenWithNoConfigAtAll() {
        paths.restoreOldDir("task-1").mkdirs()
        File(paths.restoreOldDir("task-1"), "metadata.json").writeText("{}")

        val scan = MaintenanceScan.scan(paths)
        assertEquals(1, scan.artifacts.size)
        assertTrue(scan.artifacts.single().hasOld)
        assertFalse(scan.artifacts.single().hasStaging)

        val reservation = MaintenanceReservation.resolve(StartupConfig.DEFAULTS, scan)
        assertTrue(reservation.required)
        assertTrue(reservation.recoveryRequired)
        assertEquals("task-1", reservation.taskId)
    }

    @Test
    fun restoreArtifactsReserveMaintenanceEvenWithACorruptConfig() {
        paths.startupConfigFile.parentFile?.mkdirs()
        paths.startupConfigFile.writeText("{ this is not json")
        paths.restoreStagingDir("task-9").mkdirs()
        File(paths.restoreStagingDir("task-9"), "archive.part").writeText("x")

        // Reading the config yields defaults with no tasks…
        val config = StartupConfig.read(paths)
        assertTrue(config.pendingTasks.isEmpty())

        // …yet the filesystem evidence still holds the reservation.
        val reservation = MaintenanceReservation.resolve(config, MaintenanceScan.scan(paths))
        assertTrue(reservation.required)
        assertEquals("task-9", reservation.taskId)
        assertTrue(reservation.reason.contains("durable"))
    }

    @Test
    fun anEmptyRestoreWorkspaceIsNotEvidenceOnItsOwn() {
        paths.restoreWorkspaceDir("task-1").mkdirs()

        val scan = MaintenanceScan.scan(paths)
        assertTrue(scan.artifacts.single().isEmpty)
        assertFalse(scan.hasDurableEvidence)
    }

    @Test
    fun aQueuedTaskReservesMaintenanceWithoutDemandingRecovery() {
        val reservation = MaintenanceReservation.resolve(
            configWith("task-1" to MaintenanceTaskState.QUEUED),
            MaintenanceScan.scan(paths),
        )

        assertTrue(reservation.required)
        assertFalse(reservation.recoveryRequired)
        assertEquals("task-1", reservation.taskId)
        assertTrue(reservation.reason.contains("queued"))
    }

    @Test
    fun aTaskLeftRunningDemandsRecoveryEvenWithACleanWorkspace() {
        val reservation = MaintenanceReservation.resolve(
            configWith("task-1" to MaintenanceTaskState.RUNNING),
            MaintenanceScan.scan(paths),
        )

        assertTrue(reservation.required)
        assertTrue(reservation.recoveryRequired)
    }

    @Test
    fun theMidOperationTaskIsPreferredOverAnEarlierQueuedOne() {
        writeJournal("task-2", MaintenanceJournalKind.RESTORE.id, RestorePhase.INSTALLED.id)

        val reservation = MaintenanceReservation.resolve(
            configWith(
                "task-1" to MaintenanceTaskState.QUEUED,
                "task-2" to MaintenanceTaskState.RUNNING,
            ),
            MaintenanceScan.scan(paths),
        )

        assertEquals("task-2", reservation.taskId)
    }
}

File: apps/mellow/test/core/startup/maintenance_scanner_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:path/path.dart' as p;
import 'package:weblibre/core/startup/maintenance_scanner.dart';
import 'package:weblibre/core/startup/models/maintenance_journal.dart';
import 'package:weblibre/core/startup/models/startup_config.dart';
import 'package:weblibre/core/startup/startup_config_store.dart';
import 'package:weblibre/core/startup/startup_paths.dart';

const _profileId = '0199a0b1-1111-7111-8111-111111111111';

void main() {
  _tolerantParsingTests();

  late Directory filesDir;
  late StartupPaths paths;
  late MaintenanceScanner scanner;

  setUp(() async {
    filesDir = await Directory.systemTemp.createTemp('weblibre_maintenance');
    paths = StartupPaths(filesDir);
    await paths.ensureGlobalDirectories();
    scanner = MaintenanceScanner(paths);
  });

  tearDown(() async {
    if (filesDir.existsSync()) {
      await filesDir.delete(recursive: true);
    }
  });

  Future<void> writeJournal(
    String taskId, {
    required MaintenanceJournalKind kind,
    required String phase,
  }) async {
    final journal = MaintenanceJournal(
      taskId: taskId,
      kindId: kind.name,
      phaseId: phase,
      targetProfileId: _profileId,
      updatedAt: DateTime.utc(2026, 8, 18),
    );
    await paths.journalFile(taskId).writeAsString(jsonEncode(journal.toJson()));
  }

  MaintenanceTask queuedTask(
    String id, {
    MaintenanceTaskState state = MaintenanceTaskState.queued,
  }) {
    return MaintenanceTask.create(
      id: id,
      action: MaintenanceAction.backup,
      state: state,
      profileId: _profileId,
      profileName: 'Default',
      createdAt: DateTime.utc(2026, 8, 18),
    );
  }

  test('an empty workspace produces no reservation', () async {
    final scan = await scanner.scan();

    expect(scan.hasDurableEvidence, isFalse);
    final reservation = MaintenanceReservation.resolve(
      StartupConfig.defaults,
      scan,
    );
    expect(reservation.required, isFalse);
    expect(reservation.recoveryRequired, isFalse);
  });

  test('a completed journal alone does not reserve maintenance', () async {
    await writeJournal(
      'task-1',
      kind: MaintenanceJournalKind.delete,
      phase: DeletePhase.completed.name,
    );

    final scan = await scanner.scan();
    expect(scan.journals, hasLength(1));
    expect(scan.incompleteJournals, isEmpty);
    expect(scan.hasDurableEvidence, isFalse);
  });

  test(
    'an incomplete journal reserves maintenance and demands recovery',
    () async {
      await writeJournal(
        'task-1',
        kind: MaintenanceJournalKind.restore,
        phase: RestorePhase.oldMoved.name,
      );

      final reservation = MaintenanceReservation.resolve(
        StartupConfig.defaults,
        await scanner.scan(),
      );

      expect(reservation.required, isTrue);
      expect(reservation.recoveryRequired, isTrue);
      expect(reservation.taskId, 'task-1');
      expect(reservation.scan.profilesUnderMaintenance, {_profileId});
    },
  );

  test('an unreadable journal is reported, not quarantined away', () async {
    await paths.journalFile('task-1').writeAsString('{ not json');

    final scan = await scanner.scan();

    expect(scan.unreadableJournals, hasLength(1));
    expect(scan.hasDurableEvidence, isTrue);
    // The evidence must survive the scan, or the next process would boot the
    // possibly half-restored profile.
    expect(paths.journalFile('task-1').existsSync(), isTrue);
  });

  test('a journal without identifying fields counts as unreadable', () async {
    await paths.journalFile('task-1').writeAsString('{"phase":"staged"}');

    final scan = await scanner.scan();
    expect(scan.unreadableJournals, hasLength(1));
    expect(scan.journals, isEmpty);
  });

  test(
    'restore artifacts reserve maintenance even with no config at all',
    () async {
      await paths.restoreOldDir('task-1').create(recursive: true);
      await File(
        p.join(paths.restoreOldDir('task-1').path, 'metadata.json'),
      ).writeAsString('{}');

      final scan = await scanner.scan();
      expect(scan.artifacts, hasLength(1));
      expect(scan.artifacts.single.hasOld, isTrue);
      expect(scan.artifacts.single.hasStaging, isFalse);

      final reservation = MaintenanceReservation.resolve(
        StartupConfig.defaults,
        scan,
      );
      expect(reservation.required, isTrue);
      expect(reservation.recoveryRequired, isTrue);
      expect(reservation.taskId, 'task-1');
    },
  );

  test(
    'restore artifacts reserve maintenance even with a corrupt config',
    () async {
      await paths.startupConfigFile.writeAsString('{ this is not json');
      await paths.restoreStagingDir('task-9').create(recursive: true);
      await File(
        p.join(paths.restoreStagingDir('task-9').path, 'archive.part'),
      ).writeAsString('x');

      // Reading the config quarantines it and yields defaults with no tasks…
      final config = await StartupConfigStore(paths).read();
      expect(config.pendingTasks, isEmpty);

      // …yet the filesystem evidence still holds the reservation.
      final reservation = MaintenanceReservation.resolve(
        config,
        await scanner.scan(),
      );
      expect(reservation.required, isTrue);
      expect(reservation.taskId, 'task-9');
      expect(reservation.reason, contains('durable'));
    },
  );

  test('an empty restore workspace is not evidence on its own', () async {
    await paths.restoreWorkspaceDir('task-1').create(recursive: true);

    final scan = await scanner.scan();
    expect(scan.artifacts.single.isEmpty, isTrue);
    expect(scan.hasDurableEvidence, isFalse);
  });

  test(
    'a queued task reserves maintenance without demanding recovery',
    () async {
      final config = StartupConfig(pendingTasks: [queuedTask('task-1')]);

      final reservation = MaintenanceReservation.resolve(
        config,
        await scanner.scan(),
      );

      expect(reservation.required, isTrue);
      expect(reservation.recoveryRequired, isFalse);
      expect(reservation.taskId, 'task-1');
      expect(reservation.reason, contains('queued'));
    },
  );

  test(
    'a task left running demands recovery even with a clean workspace',
    () async {
      final config = StartupConfig(
        pendingTasks: [
          queuedTask('task-1', state: MaintenanceTaskState.running),
        ],
      );

      final reservation = MaintenanceReservation.resolve(
        config,
        await scanner.scan(),
      );

      expect(reservation.required, isTrue);
      expect(reservation.recoveryRequired, isTrue);
    },
  );

  test(
    'the mid-operation task is preferred over an earlier queued one',
    () async {
      await writeJournal(
        'task-2',
        kind: MaintenanceJournalKind.restore,
        phase: RestorePhase.installed.name,
      );

      final config = StartupConfig(
        pendingTasks: [
          queuedTask('task-1'),
          queuedTask('task-2', state: MaintenanceTaskState.running),
        ],
      );

      final reservation = MaintenanceReservation.resolve(
        config,
        await scanner.scan(),
      );

      expect(reservation.taskId, 'task-2');
    },
  );
}

/// A journal is evidence that a destructive operation was in flight, and it is
/// read on the startup path before anything else can run. A field of the wrong
/// *type* — not merely missing — must therefore degrade to a default, exactly as
/// a hand-written tolerant reader did, and must never take the scan down with
/// it: an exception here is an app that cannot start until the file is deleted
/// by hand.
void _tolerantParsingTests() {
  group('malformed journal fields', () {
    late Directory filesDir;
    late StartupPaths paths;

    setUp(() async {
      filesDir = await Directory.systemTemp.createTemp('weblibre_malformed');
      paths = StartupPaths(filesDir);
      await paths.ensureGlobalDirectories();
    });

    tearDown(() async {
      if (filesDir.existsSync()) {
        await filesDir.delete(recursive: true);
      }
    });

    Map<String, Object?> wellFormed() => {
      'version': 1,
      'taskId': 'task-1',
      'kind': 'restore',
      'phase': 'staged',
      'targetProfileId': _profileId,
      'participants': <Object?>[],
      'updatedAt': DateTime.utc(2026, 8, 18).toIso8601String(),
    };

    test('a string version falls back instead of throwing', () {
      final journal = MaintenanceJournal.tryFromJson({
        ...wellFormed(),
        'version': '1',
      });

      expect(journal, isNotNull);
      expect(journal!.version, maintenanceJournalVersion);
      // Still a recognisable restore journal, not a write-off.
      expect(journal.kind, MaintenanceJournalKind.restore);
    });

    test('non-string kind and phase fall back instead of throwing', () {
      final journal = MaintenanceJournal.tryFromJson({
        ...wellFormed(),
        'kind': 1,
        'phase': false,
      });

      expect(journal, isNotNull);
      expect(journal!.kindId, '');
      expect(journal.phaseId, '');
      // Unrecognised, so the next process offers recovery rather than booting.
      expect(journal.isUnrecognised, isTrue);
    });

    test('a malformed participant version falls back instead of throwing', () {
      final journal = MaintenanceJournal.tryFromJson({
        ...wellFormed(),
        'participants': [
          {'id': 'secureStorage', 'version': '2', 'state': 'prepared'},
        ],
      });

      expect(journal, isNotNull);
      expect(journal!.participants.single.version, 0);
      expect(journal.participants.single.state, ParticipantState.prepared);
    });

    test('scan reports a malformed journal instead of throwing', () async {
      await paths
          .journalFile('task-1')
          .writeAsString(
            jsonEncode({...wellFormed(), 'version': '1', 'kind': 7}),
          );

      final scan = await MaintenanceScanner(paths).scan();

      expect(scan.journals, hasLength(1));
      expect(scan.unreadableJournals, isEmpty);
      expect(scan.journals.single.taskId, 'task-1');
    });
  });
}
